File: 0525/dp.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
  int findMaxLength(vector<int>& nums) {
    unordered_map<int, int> map;
    int balance = 0;
    int ret = 0;
    map[0] = -1;
    for (int i = 0; i < nums.size(); ++i) {
      if (nums[i] == 1) balance++;
      else balance--;
      if (map.find(balance) != map.end()) {
        ret = max(ret, i - map[balance]);
      } else {
        map[balance] = i;
      }
    }
    return ret;
  }
};

int main() {
  Solution sol;
  vector<int> nums;

  nums = {0,1};
  assert(sol.findMaxLength(nums) == 2);

  nums = {0,1,0};
  assert(sol.findMaxLength(nums) == 2);

  nums = {0,0,1,0,0};
  assert(sol.findMaxLength(nums) == 2);

  nums = {0,0,1,0,0,1,1,1};
  assert(sol.findMaxLength(nums) == 8);

  nums = {1,1,1,0,1,0,0,1};
  assert(sol.findMaxLength(nums) == 6);

  nums = {0,1,1,0,1,1,1,0};
  assert(sol.findMaxLength(nums) == 4);

  nums = {};
  assert(sol.findMaxLength(nums) == 0);
}

File: 0525/dp.rs
use std::collections::HashMap;

struct Solution;

impl Solution {
    pub fn find_max_length(nums: Vec<i32>) -> i32 {
        let mut ret: i32 = 0;
        let mut val: i32 = 0;
        let mut hashmap = HashMap::new();
        hashmap.insert(val, -1);
        for idx in 0..nums.len() {
            if nums[idx] == 1 {
                val += 1;
            } else {
                val -= 1;
            }
            hashmap.entry(val).or_insert(idx as i32);
            match hashmap.get(&val) {
                Some(&prev_idx) => { ret = std::cmp::max(ret, idx as i32 - prev_idx as i32); }
                None => {}
            }
        }
        return ret;
    }
}

fn main() {
    let mut nums: Vec<i32>;

    nums = vec![0, 1];
    assert!(Solution::find_max_length(nums) == 2);

    nums = vec![0, 1, 0];
    assert!(Solution::find_max_length(nums) == 2);

    nums = vec![];
    assert!(Solution::find_max_length(nums) == 0);

    nums = vec![0, 1, 0, 0, 1, 0];
    assert!(Solution::find_max_length(nums) == 4);
}
